scio-snowflake/src/main/scala/com/spotify/scio/snowflake/SnowflakeIO.scala

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.snowflake
+
+import com.spotify.scio.ScioContext
+import com.spotify.scio.coders.{Coder, CoderMaterializer}
+import com.spotify.scio.io.{EmptyTap, EmptyTapOf, ScioIO, Tap, TapT}
+import com.spotify.scio.values.SCollection
+import kantan.csv.{RowDecoder, RowEncoder}
+import org.apache.beam.sdk.io.snowflake.SnowflakeIO.{CsvMapper, UserDataMapper}
+import org.apache.beam.sdk.io.{snowflake => beam}
+
+object SnowflakeIO {
+
+  private[snowflake] def dataSourceConfiguration(connectionOptions: SnowflakeConnectionOptions) = {
+
+    val datasourceInitial = beam.SnowflakeIO.DataSourceConfiguration
+      .create()
+
+    val datasourceWithAuthent = connectionOptions.authenticationOptions match {
+      case SnowflakeUsernamePasswordAuthenticationOptions(username, password) =>
+        datasourceInitial.withUsernamePasswordAuth(username, password)
+      case SnowflakeKeyPairAuthenticationOptions(username, privateKeyPath, None) =>
+        datasourceInitial.withKeyPairPathAuth(username, privateKeyPath)
+      case SnowflakeKeyPairAuthenticationOptions(username, privateKeyPath, Some(passphrase)) =>
+        datasourceInitial.withKeyPairPathAuth(username, privateKeyPath, passphrase)
+      case SnowflakeOAuthTokenAuthenticationOptions(token) =>
+        datasourceInitial.withOAuth(token)
+    }
+
+    val datasourceBeforeSchema = datasourceWithAuthent
+      .withServerName(connectionOptions.serverName)
+      .withDatabase(connectionOptions.database)
+      .withRole(connectionOptions.role)
+      .withWarehouse(connectionOptions.warehouse)
+
+    connectionOptions.schema
+      .map(schema => datasourceBeforeSchema.withSchema(schema))
+      .getOrElse(datasourceBeforeSchema)
```

Comment:
Nit: to avoid using intermediate variable, you can use syntactic sugar from `scala.util.chaining._`.
(can be done above too)

```suggestion
      .withWarehouse(connectionOptions.warehouse)
      .pipe(ds => connectionOptions.schema.fold(ds)(ds.withSchema))
```

Reply:
With `fold`: a9b56c4333a33dfad326a71441634ac23c37c0e9